homeworks/solutions/krasnor/2-ddg/src/halfedge.js

```diff
@@ -0,0 +1,634 @@
+                faces.set(e.face.idx, e.face);
+        }
+        return faces.values();
+    }
+
+    /**
+     * Returns a Map of all Halfedges that point away from this vertex. Key is the corresponding face index
+     *
+     * @returns {Map<Number, Halfedge>}
+     */
+    getAllOutgoingHalfedges() {
+        const e0 = this.halfedge;
+        let hfedges = new Map();
+        if (e0.vertex.idx == this.idx)
+            hfedges.set(e0.face.idx, e0);
+
+        for (let e = e0.twin.next; e != e0; e = e.twin.next) {
+            if (e == null || e == undefined) {
+                break;
+            }
+            if (e.vertex.idx == this.idx)
+                hfedges.set(e.face.idx, e);
+        }
+        return hfedges;
+    }
+
+    /**
+     * another way of halfedge traversal. used for cross checking code
+     * @param fn
+     */
+    halfedges(fn) { // given vertex
+        let start = true
+        let i = 0
+        for (let h = this.halfedge; start || h != this.halfedge; h = h.twin.next) {
+            if (h == null || h == undefined) // not part of an face -> boundary
+                break;
+
+            fn(h, i)
+            start = false
+            i++
+        }
+    }
+}
+
+export class HalfedgeMesh {
+    /**
+     * constructor constructs the halfedge-based mesh representation.
+     *
+     * @param {string} data is a text string from an .obj file
+     */
+    constructor(data) {
+        // properties we plan to cache
+        this.vertices = [] // an array of Vertex object
+        this.edges = [] // an array of Edge object
+        this.faces = [] // an array of Face object
+        this.halfedges = [] // an array of Halfedge object
+
+        this.edgeIdCache = new Map(); // Map<string, edgeId> key: "v0|v1"
+
+        console.log('Start creating half edge - ' + new Date().toTimeString());
+        console.log("This may take about 2-5 seconds"); // next performance gain would be avoiding regex
+
+        let start = new Date().getTime();
+        let lines = data.split('\n');
+
+        // 1. pass parse verts and faces
+        for (let line of lines) {
+            line = line.trim();
+            const tokens = line.split(' ');
+            switch (tokens[0].trim()) {
+                case 'v':
+                    let vPos = new Vector(
+                        parseFloat(tokens[1]),
+                        parseFloat(tokens[2]),
+                        parseFloat(tokens[3]),
+                        1,
+                    );
+                    this.vertices.push(new Vertex(vPos, this.vertices.length));
+                    // halfedge will be set when creating faces
+                    continue
+                case 'vt':
+                    // uvs not needed in this implementation task
+                    continue
+                case 'vn':
+                    // normals not needed in this implementation task
+                    continue
+                case 'f':
```

Comment:
It is quite impressive that you deal with the construction while parsing phase.

For our provided model, your approach works. But just keep in mind, a nit issue: faces are not guaranteed to appear after all vertices are loaded. Moreover, greedy construction might increase the difficulty of handling twin half edges (In your case you always create the twin when creating a halfedge), because we may also interested in the mesh boundaries.